File: tools/config-tui/src/lib/schema-to-fields.ts
import { type FormField } from "../components/form.js";

interface SchemaProperty {
  type?: string;
  description?: string;
  enum?: string[];
  pattern?: string;
  minimum?: number;
  maximum?: number;
  minLength?: number;
  maxLength?: number;
  properties?: Record<string, SchemaProperty>;
  required?: string[];
  additionalProperties?: boolean;
  $ref?: string;
}

interface JsonSchema {
  properties?: Record<string, SchemaProperty>;
  required?: string[];
  $defs?: Record<string, SchemaProperty>;
}

export interface FormSection {
  title: string;
  fields: FormField[];
}

/**
 * Generate sections with fields from a JSON Schema + matching data object.
 * Top-level objects become sections; leaf properties become fields.
 */
export function schemaToSections(
  schema: JsonSchema,
  data: Record<string, unknown>,
): FormSection[] {
  const sections: FormSection[] = [];
  const props = schema.properties ?? {};
  const defs = schema.$defs ?? {};

  for (const [key, prop] of Object.entries(props)) {
    const resolved = resolveRef(prop, defs);
    const value = data[key];

    if (resolved.type === "object" && resolved.properties) {
      // This becomes a section
      const fields = objectToFields(
        resolved.properties,
        defs,
        (value as Record<string, unknown>) ?? {},
        key,
      );
      sections.push({ title: formatLabel(key), fields });
    } else {
      // Top-level non-object fields go into a "General" section
      const existing = sections.find((s) => s.title === "General");
      const field = propertyToField(`${key}`, key, resolved, value);
      if (existing) {
        existing.fields.push(field);
      } else {
        sections.unshift({ title: "General", fields: [field] });
      }
    }
  }

  return sections;
}

/**
 * Flat version for backward compat — returns FormField[] with section separators.
 */
export function schemaToFields(
  schema: JsonSchema,
  data: Record<string, unknown>,
  prefix = "",
): FormField[] {
  const sections = schemaToSections(schema, data);
  const fields: FormField[] = [];

  for (const section of sections) {
    // Insert a separator field
    fields.push({
      key: `__section__${section.title}`,
      label: section.title,
      type: "section" as any,
      value: null,
    });
    fields.push(...section.fields);
  }

  return fields;
}

function objectToFields(
  properties: Record<string, SchemaProperty>,
  defs: Record<string, SchemaProperty>,
  data: Record<string, unknown>,
  parentKey: string,
): FormField[] {
  const fields: FormField[] = [];

  for (const [key, prop] of Object.entries(properties)) {
    const resolved = resolveRef(prop, defs);
    const fullKey = `${parentKey}.${key}`;
    const value = data[key];

    if (resolved.type === "object" && resolved.properties) {
      // Nested sub-object: flatten with a sub-section separator
      fields.push({
        key: `__section__${fullKey}`,
        label: formatLabel(key),
        type: "section" as any,
        value: null,
      });
      fields.push(
        ...objectToFields(resolved.properties, defs, (value as Record<string, unknown>) ?? {}, fullKey),
      );
    } else {
      fields.push(propertyToField(fullKey, key, resolved, value));
    }
  }

  return fields;
}

/**
 * Reconstruct a nested object from flat dot-key values.
 */
export function fieldsToData(values: Record<string, unknown>): Record<string, unknown> {
  const result: Record<string, unknown> = {};

  for (const [dotKey, val] of Object.entries(values)) {
    if (dotKey.startsWith("__section__")) continue;
    const parts = dotKey.split(".");
    let target: Record<string, unknown> = result;
    for (let i = 0; i < parts.length - 1; i++) {
      if (!(parts[i] in target)) target[parts[i]] = {};
      target = target[parts[i]] as Record<string, unknown>;
    }
    target[parts[parts.length - 1]] = val;
  }

  return result;
}

function resolveRef(prop: SchemaProperty, defs: Record<string, SchemaProperty>): SchemaProperty {
  if (!prop.$ref) return prop;
  const refName = prop.$ref.replace("#/$defs/", "");
  const resolved = defs[refName];
  if (!resolved) return prop;
  return resolveRef(resolved, defs);
}

function propertyToField(fullKey: string, rawKey: string, prop: SchemaProperty, value: unknown): FormField {
  const label = formatLabel(rawKey);

  if (prop.enum) {
    return { key: fullKey, label, type: "enum", value: value ?? prop.enum[0], enumValues: prop.enum, description: prop.description };
  }
  if (prop.type === "boolean") {
    return { key: fullKey, label, type: "boolean", value: value ?? false, description: prop.description };
  }
  if (prop.type === "integer") {
    return { key: fullKey, label, type: "integer", value: value ?? 0, min: prop.minimum, max: prop.maximum, description: prop.description };
  }
  if (prop.type === "string" && prop.pattern === "^#[0-9A-Fa-f]{6}$") {
    return { key: fullKey, label, type: "color", value: value ?? "#000000", pattern: prop.pattern, description: prop.description };
  }
  return { key: fullKey, label, type: "string", value: value ?? "", pattern: prop.pattern, description: prop.description };
}

function formatLabel(key: string): string {
  return key
    .replace(/([a-z])([A-Z])/g, "$1 $2")
    .replace(/-/g, " ")
    .replace(/\b\w/g, (c) => c.toUpperCase());
}

File: tools/config-tui-py/vonage_config_tui/schema_to_fields.py
"""JSON Schema → flat list of FormField."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any, Optional


@dataclass
class FormField:
    key: str
    label: str
    type: str  # "boolean" | "string" | "enum" | "integer" | "color" | "section"
    value: Any = None
    enum_values: Optional[list[str]] = None
    description: Optional[str] = None
    min: Optional[int] = None
    max: Optional[int] = None
    pattern: Optional[str] = None


_CAMEL_SPLIT = re.compile(r"([a-z])([A-Z])")

_HEX_COLOR_PATTERNS = {
    "^#[0-9A-Fa-f]{6}$",
    "^#([0-9A-Fa-f]{6}|[0-9A-Fa-f]{8})$",
}


def _is_hex_color_pattern(pattern: Optional[str]) -> bool:
    return pattern in _HEX_COLOR_PATTERNS


def _format_label(key: str) -> str:
    result = _CAMEL_SPLIT.sub(r"\1 \2", key).replace("-", " ")
    return " ".join(w[:1].upper() + w[1:] for w in result.split(" ") if w)


def _resolve_ref(prop: dict[str, Any], defs: dict[str, Any]) -> dict[str, Any]:
    seen: set[str] = set()
    while isinstance(prop, dict) and "$ref" in prop:
        ref = prop["$ref"]
        if ref in seen:
            return prop
        seen.add(ref)
        name = ref.replace("#/$defs/", "")
        resolved = defs.get(name)
        if not resolved:
            return prop
        prop = resolved
    return prop


def _property_to_field(full_key: str, raw_key: str, prop: dict[str, Any], value: Any) -> FormField:
    label = _format_label(raw_key)
    desc = prop.get("description")

    if "enum" in prop:
        enum_vals = list(prop["enum"])
        return FormField(
            key=full_key,
            label=label,
            type="enum",
            value=value if value is not None else enum_vals[0],
            enum_values=enum_vals,
            description=desc,
        )
    if prop.get("type") == "boolean":
        return FormField(
            key=full_key,
            label=label,
            type="boolean",
            value=bool(value) if value is not None else False,
            description=desc,
        )
    if prop.get("type") == "integer":
        return FormField(
            key=full_key,
            label=label,
            type="integer",
            value=value if value is not None else 0,
            min=prop.get("minimum"),
            max=prop.get("maximum"),
            description=desc,
        )
    if prop.get("type") == "number":
        return FormField(
            key=full_key,
            label=label,
            type="integer",
            value=value if value is not None else 0,
            min=prop.get("minimum"),
            max=prop.get("maximum"),
            description=desc,
        )
    if prop.get("type") == "string" and _is_hex_color_pattern(prop.get("pattern")):
        return FormField(
            key=full_key,
            label=label,
            type="color",
            value=value if value is not None else "#000000",
            pattern=prop.get("pattern"),
            description=desc,
        )
    return FormField(
        key=full_key,
        label=label,
        type="string",
        value=value if value is not None else "",
        pattern=prop.get("pattern"),
        description=desc,
    )


def _object_to_fields(
    properties: dict[str, Any],
    defs: dict[str, Any],
    data: dict[str, Any],
    parent_key: str,
) -> list[FormField]:
    fields: list[FormField] = []
    for key, prop in properties.items():
        resolved = _resolve_ref(prop, defs)
        full_key = f"{parent_key}.{key}"
        value = data.get(key) if isinstance(data, dict) else None
        if resolved.get("type") == "object" and "properties" in resolved:
            fields.append(FormField(
                key=f"__section__{full_key}",
                label=_format_label(key),
                type="section",
            ))
            fields.extend(
                _object_to_fields(resolved["properties"], defs, value or {}, full_key)
            )
        else:
            fields.append(_property_to_field(full_key, key, resolved, value))
    return fields


def schema_to_fields(schema: dict[str, Any], data: dict[str, Any]) -> list[FormField]:
    """Flatten a schema into an ordered list of fields with 'section' separators.

    Optional object properties (not in the schema's ``required`` list) that are absent
    from ``data`` are skipped entirely, rather than materialized as an empty section —
    otherwise saving would write out an empty object that fails the sub-schema's own
    ``required`` fields (e.g. platform-specific sections like ``localizationSettings``
    that Android's app-config.json never populates).
    """
    props = schema.get("properties", {}) or {}
    defs = schema.get("$defs", {}) or {}
    required = set(schema.get("required", []) or [])

    general: list[FormField] = []
    sections: list[tuple[str, list[FormField]]] = []

    for key, prop in props.items():
        resolved = _resolve_ref(prop, defs)
        is_present = isinstance(data, dict) and key in data
        if resolved.get("type") == "object" and "properties" in resolved:
            if key not in required and not is_present:
                continue
            value = data.get(key) if isinstance(data, dict) else None
            section_fields = _object_to_fields(resolved["properties"], defs, value or {}, key)
            sections.append((_format_label(key), section_fields))
        else:
            value = data.get(key) if isinstance(data, dict) else None
            general.append(_property_to_field(key, key, resolved, value))

    out: list[FormField] = []
    if general:
        out.append(FormField(key="__section__General", label="General", type="section"))
        out.extend(general)
    for title, section_fields in sections:
        out.append(FormField(key=f"__section__{title}", label=title, type="section"))
        out.extend(section_fields)
    return out


def fields_to_data(values: dict[str, Any]) -> dict[str, Any]:
    """Reconstruct a nested object from flat ``a.b.c`` keys."""
    result: dict[str, Any] = {}
    for dot_key, val in values.items():
        if dot_key.startswith("__section__"):
            continue
        parts = dot_key.split(".")
        target: dict[str, Any] = result
        for p in parts[:-1]:
            existing = target.get(p)
            if not isinstance(existing, dict):
                target[p] = {}
            target = target[p]  # type: ignore[assignment]
        target[parts[-1]] = val
    return result
